File: src/BinaryTree/Exercise105_2.java
package BinaryTree;

import java.util.HashMap;

import BinaryTree.Exercise100.TreeNode;

public class Exercise105_2 {
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;

        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    class Solution {
        public TreeNode buildTree(int[] preorder, int[] inorder) {
            HashMap<Integer, Integer> inMap = new HashMap<>();
            for (int i = 0; i < inorder.length; i++)
                inMap.put(inorder[i], i);
            TreeNode root = buildTree(preorder, 0, preorder.length - 1, inorder, 0, inorder.length - 1, inMap);
            return root;
        }

        public TreeNode buildTree(int[] preorder, int preStart, int preEnd, int[] inorder, int inStart, int inEnd,
                HashMap<Integer, Integer> inMap) {
            if (preStart > preEnd || inStart > inEnd)
                return null;

            TreeNode root = new TreeNode(preorder[preStart]);
            int inRoot = inMap.get(root.val);
            int numsLeft = inRoot - inStart;

            root.left = buildTree(preorder, preStart + 1, preStart + numsLeft, inorder, inStart, inRoot - 1, inMap);
            root.right = buildTree(preorder, preStart + numsLeft + 1, preEnd, inorder, inRoot + 1, inEnd, inMap);
            return root;
        }

    }

    public static void main(String[] args) {
        Exercise105_2 exercise = new Exercise105_2();
        Solution solution = exercise.new Solution();
        /*
         * TreeNode node8 = exercise.new TreeNode(8);
         * TreeNode node7 = exercise.new TreeNode(7, null, node8);
         * TreeNode node6 = exercise.new TreeNode(6);
         * TreeNode node5 = exercise.new TreeNode(5, null, node7);
         * TreeNode node4 = exercise.new TreeNode(4, node5, node6);
         * TreeNode node3 = exercise.new TreeNode(20,node4,node5);
         * TreeNode node2 = exercise.new TreeNode(9);
         * TreeNode node1 = exercise.new TreeNode(3,node2, node3);
         */
        int[] preorder = new int[] { 3, 9, 20, 15, 7 };
        int[] inorder = new int[] { 9, 3, 15, 20, 7 };
        TreeNode root = solution.buildTree(preorder, inorder);
    }
}

File: src/BinaryTree/Exercise105_2.cpp
/* Note:not solved by myself, pay attention to the index */
#include <iostream>
#include "BinaryTree.h"
#include <unordered_map>
using namespace std;

class Solution
{
public:
    TreeNode *buildTree(vector<int> &preorder, vector<int> &inorder)
    {
        unordered_map<int, int> inorderMap;
        for (int i = 0; i < preorder.size(); i++)
        {
            inorderMap.insert({inorder[i], i});
        }
        return buildTree(preorder, 0, (int)preorder.size() - 1, inorderMap);
    }

private:
    int index = 0;

    TreeNode *buildTree(vector<int> &preorder, int start, int end, unordered_map<int, int> &inorderMap)
    {

        if (start > end)
            return nullptr;

        int rootVal = preorder[index++];
        TreeNode *root = new TreeNode(rootVal);
        int mid = inorderMap[rootVal];
        root->left = buildTree(preorder, start, mid - 1, inorderMap);
        root->right = buildTree(preorder, mid + 1, end, inorderMap);
        return root;
    }
};

int main()
{
    vector<int> preorder{1, 2, 4, 8, 9, 5, 10, 11, 3, 6, 12, 13, 7, 14, 15};
    vector<int> inorder{8, 4, 9, 2, 10, 5, 11, 1, 12, 6, 13, 3, 14, 7, 15};
    Solution sl;
    TreeNode *root = sl.buildTree(preorder, inorder);
    root->printTree(root);
}